src/ansi.rs

```diff
@@ -48,17 +48,440 @@
+                    None => default,
+                    Some(n) => N::from_str(n).unwrap(),
+                },
+            )
+        })
+    }
+}
+
+/// Parse x ; y
+fn dual_int_sequence_atom<N: FromStr>(input: &str) -> IResult<&str, (N, N)>
+where
+    <N as FromStr>::Err: Debug,
+{
+    nom::sequence::separated_pair(
+        nom::character::streaming::digit1,
+        nom::character::streaming::char(';'),
+        nom::character::streaming::digit1,
+    )(input)
+    .map(|(rest, (a, b))| (rest, (N::from_str(a).unwrap(), N::from_str(b).unwrap())))
+}
+
+// This will parse x ; y <end>
+fn dual_int_parameter_sequence<N: FromStr>(
+    ending: char,
+) -> impl FnMut(&str) -> IResult<&str, (N, N)>
+where
+    <N as FromStr>::Err: Debug,
+{
+    move |input: &str| {
+        let params = nom::sequence::separated_pair(
+            nom::character::streaming::digit1,
+            nom::character::streaming::char(';'),
+            nom::character::streaming::digit1,
+        );
+        sequence_with_ending(params, ending)(input)
+            .map(|(rest, (a, b))| (rest, (N::from_str(a).unwrap(), N::from_str(b).unwrap())))
+    }
+}
+
+fn cursor_to_0(input: &str) -> OpResult {
+    nom::character::streaming::char('H')(input)
+        .map(|(rest, _)| (rest, Op::MoveCursorAbs { x: 0, y: 0 }))
+}
+
+fn cursor_to_line_col(input: &str) -> OpResult {
+    // Recognize <digits> `;` <digits>
+    // Start with CSI and end with H or f
+    nom::branch::alt((
+        dual_int_parameter_sequence('H'),
+        dual_int_parameter_sequence('f'),
+    ))(input)
+    .map(|(rest, (a, b))| (rest, Op::MoveCursorAbs { x: a, y: b }))
+}
+
+/// ESC [ <n> A         => Cursor up n lines
+fn cursor_up_lines(input: &str) -> OpResult {
+    optional_int_param_sequence::<isize>('A', 1)(input)
+        .map(|(rest, n)| (rest, Op::MoveCursorDelta { dx: 0, dy: -n }))
+}
+
+/// ESC [ <n> B
+fn cursor_down_lines(input: &str) -> OpResult {
+    optional_int_param_sequence::<isize>('B', 1)(input)
+        .map(|(rest, n)| (rest, Op::MoveCursorDelta { dx: 0, dy: n }))
+}
+
+// ESC [ <n> C         => Cursor right n columns
+fn cursor_right_col(input: &str) -> OpResult {
+    optional_int_param_sequence::<isize>('C', 1)(input)
+        .map(|(rest, n)| (rest, Op::MoveCursorDelta { dx: n, dy: 0 }))
+}
+
+// ESC [ <n> D         => Cursor left n columns
+fn cursor_left_col(input: &str) -> OpResult {
+    optional_int_param_sequence::<isize>('D', 1)(input)
+        .map(|(rest, n)| (rest, Op::MoveCursorDelta { dx: -n, dy: 0 }))
+}
+
+// ESC [ <n> E         => Cursor to beginning of next line, n lines down
+fn cursor_beginning_down(input: &str) -> OpResult {
+    optional_int_param_sequence::<isize>('E', 1)(input)
+        .map(|(rest, n)| (rest, Op::MoveCursorBeginningAndLine { dy: n }))
+}
+
+// ESC [ <n> F         => Cursor to beginning of prev line, n lines up
+fn cursor_beginning_up(input: &str) -> OpResult {
+    optional_int_param_sequence::<isize>('E', 1)(input)
```

Comment:
```suggestion
    optional_int_param_sequence::<isize>('F', 1)(input)
```

This is a good example of what I mean by "obscured similarities": I'm not saying this couldn't have happened if we allowed this to be anonymous instead of needing to introduce a `cursor_beginning_up` name, but I do think it would be less likely.